floodmodeller_api/toolbox/model_build/structure_log/structure_log.py

```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
-        length = 0.0
-        inlet = ""
-        outlet = ""
+        conduit_data = {"length": conduit.dist_to_next}
+        # modified conduit crawler script
+        add_to_conduit_stack = None
+
+        # check if the previous node is an inlet
         previous = self._dat.prev(conduit)
         if hasattr(previous, "subtype") and previous.subtype == "INLET":
-            inlet = previous.ki
+            conduit_data["inlet"] = previous.ki
+
         current_conduit = conduit
-        while current_conduit.dist_to_next != 0:
-            length += current_conduit.dist_to_next
-            current_conduit = self._dat.next(current_conduit)
-        next_conduit = self._dat.next(current_conduit)
-        if hasattr(next_conduit, "subtype") and next_conduit.subtype == "OUTLET":
-            outlet = next_conduit.loss_coefficient
-        return [length, inlet, outlet]
-
-    def _culvert_loss_data(self, inlet, outlet):
-        culvert_loss = ""
-        if inlet != "" and outlet != "":
-            culvert_loss = f"Ki: {inlet}, Ko: {outlet}"
-        elif inlet != "":
-            culvert_loss = f"Ki: {inlet}"
-        elif outlet != "":
-            culvert_loss = f"Ko: {outlet}"
-        return culvert_loss
-
-    def _circular_data(self, conduit, length):
-        dimensions = f"dia: {conduit.diameter:.2f} x l: {length:.2f}"
-        all_mannings = [
+
+        if current_conduit.name not in self.already_in_chain:
+            # if this conduit isnt part of a chain already, then it must be part of a new chain
+            total_length = 0
+            chain = []
+            while True:
+                self.already_in_chain.add(current_conduit.name)
+                chain.append(current_conduit.name)
+                if current_conduit.dist_to_next == 0:
+                    break
+
+                total_length += current_conduit.dist_to_next
+                current_conduit = self._dat.next(current_conduit)
+
+            self.conduit_chains[conduit.name] = chain.copy()
+            conduit_data["total_length"] = total_length
+
+        next_conduit = self._dat.next(conduit)
+        if next_conduit is not None:
+            if hasattr(next_conduit, "subtype") and next_conduit.subtype == "OUTLET":
+                conduit_data["outlet"] = next_conduit.loss_coefficient
+
+            if next_conduit._unit == "REPLICATE":
+                # for replicates, pass down the label of the unit its copying from.
+                if current_conduit._unit == "REPLICATE":
+                    next_conduit.mimic = current_conduit.mimic
+                else:
+                    next_conduit.mimic = current_conduit.name
+
+                # Replicates arent in the DAT.structures dict so we need to add them manually.
+                add_to_conduit_stack = copy.deepcopy(next_conduit)
+
+        return conduit_data, add_to_conduit_stack
+
+    def _circular_data(self, conduit):
+        dimensions = extract_attrs(conduit, {"diameter", "invert"})
```

Comment:
It wouldn't make much of a difference here due to the size (more of a matter of general good practice) but it's more efficient to iterate through lists `["diameter", "invert"]` than through sets `{"diameter", "invert"}` - sets are better for checking membership. This happens in a few places.

Reply:
I used sets here as the entries have to be unique, although Ive not heard the difference between iterating through lists and sets before